### ISSUE TITLE
Graceful shutdown of batching audit webhook

### ISSUE THREAD

**Issue by crassirostris** [kubernetes/kubernetes#50026]
Currently, in the batching audit webhook batches are sent [concurrently without any control over the lifecycle of the sending routine](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go#L237). When apiserver is being terminated, webhook doesn't wait for all request to be terminated and it may result in some audit events being dropped. Webhook should probably wait until all requests have finished before exiting the [working routine](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/plugin/pkg/audit/webhook/webhook.go#L192)

/cc @sttts @tallclair @ericchiang @piosz @CaoShuFeng

**Comment by ericchiang** [comment]
How do other API server components respond to the API server shutting down. I don't see anything in the current config that allows us to respond to os signal or something similar.

https://github.com/kubernetes/kubernetes/blob/508ccc403209e942cf1495b70f97e8eaae1fd35f/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L257

**Comment by crassirostris** [comment]
True, in all places I could find it's just `wait.NeverStop` passed during startup. @kubernetes/sig-api-machinery-misc Is there a plan to change that?

**Comment by sttts** [comment]
We have a stop channel passed into the `Run` func. It's not used everywhere (storage is not shutdown, biting us in tests right now). This could be used in combination with signals.

**Comment by mml** [comment]
cc @caesarxuchao

**Comment by soltysh** [comment]
@deads2k any ideas here? This is hitting auditing and we'd like to get this addressed before moving audit to beta.

**Comment by sttts** [comment]
cmd/hyperkube/kube-apiserver.go:36 and cmd/kube-apiserver/apiserver.go:51 are the places to put signal handling in. The main issue though is that closing a channel is not synchronous, i.e. the main func will not wait for anybody reacting on the stop channel. We need another mechanism for that, like a shutdown callback to register with.

**Comment by deads2k** [comment]
> @deads2k any ideas here? This is hitting auditing and we'd like to get this addressed before moving audit to beta.

To my knowledge, golang doesn't have a clean exit hook anywhere.  If you want guaranteed output, you probably need to use a synchronous write.

I do remember an attempt to catch interrupts (@thockin long ago iirc), but I don't remember if worked or not.

**Comment by sttts** [comment]
@soltysh we did something for audit. Do you remember?

**Comment by soltysh** [comment]
@sttts we did tests if everything is being saved, but since the legacy audit wrote stuff immediately to disk, the problem didn't exist there.

**Comment by sttts** [comment]
In fact, I doubt that the main usecase here is about writing synchronously to disk. Instead we have to make sure batched events are sent to the audit webhook. This must be synchronously done in the shutdown via some signal handler. I am pretty sure this can be done also in Golang, just need the shutdown callback registration mechanism.

**Comment by crassirostris** [comment]
The easiest temporary way would be to close the channel passed to different parts of the system and then wait for some period of time upon receiving the SIGTERM signal

**Comment by sttts** [comment]
What is "some period"? Don't like a mechanism based on hope.

**Comment by crassirostris** [comment]
@sttts That's not an actual solution, for sure. Rather something to consider if implementing a proper solution is not available any time soon

For an actual solution, I fully support you proposal with a generic shutdown callback registration

**Comment by sttts** [comment]
Let's add `RegisterShutdownHook` to k8s.io/apiserver/pkg/server.GenericApiServer. Invoke them when `GenericApiServer.Run` returns (due to the closed stop channel). Build a signal handler into the two places at https://github.com/kubernetes/kubernetes/issues/50026#issuecomment-321234946 and wait for the `Run` go-routine to finish before leaving the signal handler.

**Comment by crassirostris** [comment]
@sttts AFAIU audit module can only affect `k8s.io/apiserver/pkg/server.Config` object through `ApplyTo` method, so a method to add a hook should be there, right?

**Comment by sttts** [comment]
@crassirostris the setup of the shutdown hook can be done by the apiserver itself. It will know that audit is enabled and needs that hook.

**Comment by crassirostris** [comment]
@sttts Do you want to add a new method (like `Close`) to the audit backend interface to call from the hook?

**Comment by sttts** [comment]
@crassirostris yes, Stop or Shutdown.

**Comment by crassirostris** [comment]
@sttts I have some thoughts, would like to hear your opinion. It's not like I'm advocating this solution exclusively :)

- this method is only needed in one backend (batching webhook), but will be implemented in all of them
- more generic, even if one of the implementation of the component needs this mechanism
  - Stop method will be required in all of them
  - apiserver has to have additional logic to add this component to the list to gracefully shut down

If there's a method to add a hook in the `Config`, an implementation can decide whether it needs it or not. Also, it doesn't require to change the existing code/interfaces

**Comment by sttts** [comment]
@crassirostris how many backends do we have? :)

Alternative: create a global shutdown hook registry and just call that on shutdown from kube-apiserver. I.e. nothing plumbed through, but via a global variable. Also not perfectly nice, but maybe ok here. Maybe this even exists as a library including signal handling.

**Comment by crassirostris** [comment]
> how many backends do we have? :)

Sure :) I was mainly thinking about the number of components that require graceful shutdown growing in the future

> create a global shutdown hook registry and just call that on shutdown from kube-apiserver

Since it's in the generic apiserver, I'm not sure how correct it is to use the global scope. Also, is signal handling inseparable from the shutdown hooks? Maybe we can have a signal handing before calling the `Run` and call hooks when `stopCh` is closed?

**Comment by sttts** [comment]
We must run the hooks in the "main go routine". Waiting on `stopCh` is not enough as the process might be killed before we had a change to shut everything down.

**Comment by crassirostris** [comment]
@sttts 

> We must run the hooks in the "main go routine". Waiting on stopCh is not enough as the process might be killed before we had a change to shut everything down.

I meant running generic apiserver in blocking mode, and running a separate gorountine to catch signals

- When signals comes, a goroutine closes a channel
- Generic API server detects a closed channel and runs all hooks
- `Run` method exists and the main goroutine exits

**Comment by crassirostris** [comment]
https://github.com/kubernetes/kubernetes/pull/50439 by @sttts and https://github.com/kubernetes/kubernetes/pull/50577 by me mitigated most of the problem

The only thing thing left to do is stopping the http handler before shutting down audit logging. I'll file a separate issue for that

Thanks, @sttts!
